File: src/main/scala/io/izzel/taboolib/module/db/sql/query/Join.java
package io.izzel.taboolib.module.db.sql.query;

import io.izzel.taboolib.util.Strings;

import java.util.List;
import java.util.StringJoiner;

/**
 * io.izzel.taboolib.module.db.sql.query.Join
 *
 * @author sky
 * @since 2021/4/16 9:07 上午
 */
public class Join {

    private final JoinType type;
    private final String from;
    private final List<JoinWhere> where;

    public Join(JoinType type, String from, List<JoinWhere> where) {
        this.type = type;
        this.from = from;
        this.where = where;
    }

    public JoinType getType() {
        return type;
    }

    public String getFrom() {
        return from;
    }

    public List<JoinWhere> getWhere() {
        return where;
    }

    public String toQuery() {
        if (where.isEmpty()) {
            return Strings.replaceWithOrder("{0} join `{1}`", type.toKey(), from);
        } else {
            StringJoiner joiner = new StringJoiner(" and ");
            for (JoinWhere where : where) {
                joiner.add(Strings.replaceWithOrder("`{0}`.`{1}` {2} `{3}`.`{4}`",
                        where.getLeftTable(),
                        where.getLeftField(),
                        where.getSymbol(),
                        where.getRightTable(),
                        where.getRightField()
                ));
            }
            return Strings.replaceWithOrder("{0} join `{1}` on {2}", type.toKey(), from, joiner.toString());
        }
    }
}

File: module/module-database-core/src/main/kotlin/taboolib/module/database/Join.kt
package taboolib.module.database

/**
 * Join 语法中所包含的信息
 *
 * @author sky
 * @since 2021/6/23 3:32 下午
 */
class Join(val joinType: JoinType, val tableName: String, val filter: Filter) : Attributes {

    /** 语句 */
    override val query: String
        get() = if (filter.isEmpty()) {
            "$joinType JOIN ${tableName.asFormattedColumnName()}"
        } else {
            "$joinType JOIN ${tableName.asFormattedColumnName()} ON ${filter.query}"
        }

    /** 占位符对应的元素 */
    override val elements: List<Any>
        get() = filter.elements
}